design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,174 @@
+       user allows to be updated in parallel, OR
+    2. It is no longer `ok-to-stop` all OSDs in the CRUSH bucket.
+5. Update OSD Deployments in parallel for the last CRUSH bucket where it was `ok-to-stop` the OSDs. 
+    - If there are more OSDs in the CRUSH bucket than allowed by the user, update OSDs in the CRUSH
+      bucket in groups as large as the user configuration allows until all OSDs in the CRUSH bucket 
+      are updated.
+    - When an OSD is updated successfully, remove it from the update queue.
+7. Return to step 2 until all OSDs are updated.
+
+### How to make sure Ceph cluster updates are not starved by OSD updates
+Because [cluster growth takes precedence over updates](#cluster-growth-takes-precedence-over-updates),
+it could take a long time for all OSDs in a cluster to be updated. In order for Rook to have 
+opportunity to reconcile other components of a Ceph cluster's `CephCluster` resource, Rook should
+ensure that the OSD update reconciliation does not create a scenario where the `CephCluster` cannot
+be modified in other ways.
+
+We can achieve this by ending the OSD reconcile and requeueing the `CephCluster` reconcile after 
+the OSD updates are not able to make progress for a while. When an OSD cannot be updated 
+individually or when updating the Deployment fails, the update queue entry for that OSD should be
+updated to note that it failed. When all OSDs in the update queue queue have failed more than 3 
+times, we can regard the OSD update as having not made useful progress recently. With this, requeue
+the `CephCluster` reconcile after moving on from OSD updates.
+
+We can also establish a timeout for OSD reconciliation in order to protect against the case where
+OSD updates are making consistent progress but the updates are still taking a very long time. This
+could happen when a Ceph cluster has an extremely large number of OSDs; Rook might be able to make
+consistent progress updating the OSDs but still take several hours. In this case, we want to
+regularly make sure to allow for components other than OSDs have an opportunity to have their
+configuration updated. 15 minutes should be a good default value for this timeout, but it should
+also be user configurable. 
+
+We can actually change the previous timeout definition as an optimization. Beyond the timeout value
+Rook only needs to pause OSD updates and requeue the `CephCluster` update when there is another
+`CephCluster` update pending in Rook's queue. This can be omitted from the the first implementation
+of this feature and added later.
+
+### How to build the update queue
+List all OSD Deployments belonging to the Rook cluster to use as the update queue. Ignore OSD 
+Deployments which are already updated. OSD Deployments which need updated will have a 
+`rook-version` label that does not match the current version of the Rook operator.
+
+Go channels are likely not the best option for the update queue since it may be necessary to select
```

Comment:
Seems like we could stick with the same single-threaded approach we have currently. We have a single update queue from which we periodically extract OSDs and update their pod specs at the same time, then we wait for an event that those pods are running. Then we loop to analyze the next batch of OSDs to be updated, or create new OSDs if needed before the next batch of updates.

Reply:
That is my thought. If we can do it in a way that is single-threaded, I think that is best. I think we could have one goroutine for creating new OSDs and then use the main goroutine to update existing OSDs if it's possible for those to happen simultaneously. But the queue will just be for updates.

Reply:
For the update queue, we could use semaphores and channels for synchronization mechanisms. The queue can be timed-out if we are past the update reconcile cancellation time. So we can timeout and cancel as well if needed.

Reply:
I think we would need a semaphore/channel for indicating to the main process when progress isn't being made. But the update queue should not have any bearing when creating new OSDs.